File: codebase/backend/src/main/java/it/unisa/scanapp/domain/Transaction.java
package it.unisa.scanapp.domain;
import com.fasterxml.jackson.annotation.JsonIgnoreProperties;

import javax.persistence.*;
import javax.validation.constraints.*;

import java.io.Serializable;
import java.time.LocalDate;
import java.util.HashSet;
import java.util.Set;

/**
 * A Transaction.
 */
@Entity
@Table(name = "jhi_transx")
public class Transaction implements Serializable {

    private static final long serialVersionUID = 1L;

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    @SequenceGenerator(name = "sequenceGenerator")
    private Long id;

    @NotNull
    @Column(name = "jhi_date", nullable = false)
    private LocalDate date;

    @ManyToOne
    @JsonIgnoreProperties("transactions")
    private User user;

    @ManyToMany
    @JoinTable(name = "jhi_transx_product",
               joinColumns = @JoinColumn(name = "transaction_id", referencedColumnName = "id"),
               inverseJoinColumns = @JoinColumn(name = "product_id", referencedColumnName = "id"))
    private Set<Product> products = new HashSet<>();

    // jhipster-needle-entity-add-field - JHipster will add fields here, do not remove
    public Long getId() {
        return id;
    }

    public void setId(Long id) {
        this.id = id;
    }

    public LocalDate getDate() {
        return date;
    }

    public Transaction date(LocalDate date) {
        this.date = date;
        return this;
    }

    public void setDate(LocalDate date) {
        this.date = date;
    }

    public User getUser() {
        return user;
    }

    public Transaction user(User user) {
        this.user = user;
        return this;
    }

    public void setUser(User user) {
        this.user = user;
    }

    public Set<Product> getProducts() {
        return products;
    }

    public Transaction products(Set<Product> products) {
        this.products = products;
        return this;
    }

    public Transaction addProduct(Product product) {
        this.products.add(product);
        product.getTransactions().add(this);
        return this;
    }

    public Transaction removeProduct(Product product) {
        this.products.remove(product);
        product.getTransactions().remove(this);
        return this;
    }

    public void setProducts(Set<Product> products) {
        this.products = products;
    }
    // jhipster-needle-entity-add-getters-setters - JHipster will add getters and setters here, do not remove

    @Override
    public boolean equals(Object o) {
        if (this == o) {
            return true;
        }
        if (!(o instanceof Transaction)) {
            return false;
        }
        return id != null && id.equals(((Transaction) o).id);
    }

    @Override
    public int hashCode() {
        return 31;
    }

    @Override
    public String toString() {
        return "Transaction{" +
            "id=" + getId() +
            ", date='" + getDate() + "'" +
            "}";
    }
}

File: codebase/frontend/src/app/core/model/transaction.ts
import { Product, IProduct } from './product';
import { IUser, User } from './user';

export interface ITransaction {
    id: number;
    date: Date;
    user: IUser;
    products: IProduct[];
}

export class Transaction {
    private id: number;
    private date: Date;
    private user: User;
    private products: Array<Product>;

    constructor(transaction?: ITransaction) {
        if (transaction) {
            this.id = transaction.id;
            this.date = transaction.date;
            this.user = new User(transaction.user);
            if (transaction.products != null) {
                transaction.products.forEach(element => {
                    this.products.push(new Product(element));
                });
            }
        }
    }

    public getId(): number {
        return this.id;
    }

    public setId(id: number): void {
        this.id = id;
    }

    public getDate(): Date {
        return this.date;
    }

    public setDate(date: Date): void {
        this.date = date;
    }
    
    public getUser(): User {
        return this.user;
    }

    public setUser(user: User): void {
        this.user = user;
    }
}
